internal/model/user_reference.go

```diff
@@ -0,0 +1,189 @@
+// Inputs are parsed without checking permissions.
+// APIs using this function should require a sufficient role or custom permissions checking.
+// Returns a reference and user errors.
+// User-level errors return (partial reference, user errors).
+func ParseCourseUserReferences(rawReferences []CourseUserReference) (*ParsedCourseUserReference, map[string]error) {
+	courseUserReference := ParsedCourseUserReference{
+		Emails:                 make(map[string]any, 0),
+		ExcludeEmails:          make(map[string]any, 0),
+		CourseUserRoles:        make(map[CourseUserRole]any, 0),
+		ExcludeCourseUserRoles: make(map[CourseUserRole]any, 0),
+	}
+
+	userErrors := make(map[string]error, 0)
+
+	commonCourseRoles := GetCommonCourseUserRolesCopy()
+
+	for _, rawReference := range rawReferences {
+		reference := strings.ToLower(strings.TrimSpace(string(rawReference)))
+
+		if reference == "" {
+			continue
+		}
+
+		exclude := false
+		if strings.HasPrefix(reference, "-") {
+			exclude = true
+
+			reference = strings.TrimPrefix(reference, "-")
+			reference = strings.TrimSpace(reference)
+		}
+
+		if strings.Contains(reference, "@") {
+			if exclude {
+				courseUserReference.ExcludeEmails[reference] = nil
+			} else {
+				courseUserReference.Emails[reference] = nil
+			}
+		} else if reference == "*" {
+			allCourseRoles := make(map[CourseUserRole]any, len(commonCourseRoles))
+			for _, role := range commonCourseRoles {
+				allCourseRoles[role] = nil
+			}
+
+			if exclude {
+				courseUserReference.ExcludeCourseUserRoles = allCourseRoles
+			} else {
+				courseUserReference.CourseUserRoles = allCourseRoles
+			}
+		} else {
+			// Target a specific course role.
+			role, ok := commonCourseRoles[reference]
+			if !ok {
+				userErrors[string(rawReference)] = fmt.Errorf("Unknown course role '%s'.", reference)
+				continue
+			}
+
+			if exclude {
+				courseUserReference.ExcludeCourseUserRoles[role] = nil
+			} else {
+				courseUserReference.CourseUserRoles[role] = nil
+			}
+		}
+	}
+
+	if len(userErrors) == 0 {
+		userErrors = nil
+	}
+
+	return &courseUserReference, userErrors
+}
+
+// Returns a sorted list of users based on the course reference.
+func ResolveCourseUsers(users map[string]*CourseUser, reference *ParsedCourseUserReference) []*CourseUser {
+	if reference == nil {
+		return nil
+	}
+
+	results := make([]*CourseUser, 0, len(users))
+
+	for email, user := range users {
+		if reference.RefersTo(email, user.Role) {
+			results = append(results, user)
+		}
+	}
+
+	slices.SortFunc(results, CompareCourseUserPointer)
+
+	return results
+}
+
+// Returns a sorted list of emails based on the course reference.
+// Emails can target users outside of the course.
+func ResolveCourseUserEmails(users map[string]*CourseUser, reference *ParsedCourseUserReference) []string {
+	if reference == nil {
+		return nil
+	}
+
+	emailSet := make(map[string]any, 0)
+	excludeSet := make(map[string]any, 0)
```

Comment:
The semantics for the exclude set confuse me.

Let's say we have someone observing a course with an `other` role.
Wouldn't ['admin', 'owner', 'some-observer@test.edulinq.org'], exclude the observer?

I also don't see a test case for this situations (but I could have missed it).